tests/type-annotations/test.sh

```diff
@@ -0,0 +1,59 @@
+#!/bin/bash
+
+. /usr/share/beakerlib/beakerlib.sh || exit 1
+
+rlJournalStart
+    rlPhaseStartSetup
+        WITH_SYSTEM_PACKAGES="${WITH_SYSTEM_PACKAGES:-no}"
+        # TODO: `test` seems more natural, but creates 3 environments,
+        # one per available Python installation. I need to check whether
+        # to disable or take advantage of it.
+        HATCH_ENVIRONMENT="${HATCH_ENVIRONMENT:-dev}"
+
+        rlLogInfo "WITH_SYSTEM_PACKAGES=$WITH_SYSTEM_PACKAGES"
+        rlLogInfo "HATCH_ENVIRONMENT=$HATCH_ENVIRONMENT"
+
+        rlLogInfo "pip is $(which pip), $(pip --version)"
+        rlLogInfo "hatch is $(which hatch), $(hatch --version)"
+    rlPhaseEnd
+
+    if [ "$WITH_SYSTEM_PACKAGES" = "yes" ]; then
+        rlPhaseStartTest "Check type annotations against system Python packages"
+            rlRun "mypy_version=$(yq -r '.repos | .[] | select(.repo | test("^.*/mirrors-mypy$")) | .rev' ../../.pre-commit-config.yaml | tr -d 'v')"
+            rlRun "pyright_version=$(yq -r '.repos | .[] | select(.repo | test("^.*/pyright-python$")) | .rev' ../../.pre-commit-config.yaml | tr -d 'v')"
+
+            rlRun "rm -f requirements.txt && touch requirements.txt"
+            rlRun "echo 'mypy==v${mypy_version}' >> requirements.txt"
+            rlRun "echo 'pyright==v${pyright_version}' >> requirements.txt"
+            rlRun "yq -r '.repos | .[] | select(.repo | test(\"^.*/mirrors-mypy$\")) | .hooks[0].additional_dependencies | .[] | select(test(\"^types-.*\"))' ../../.pre-commit-config.yaml >> requirements.txt"
```

Comment:
Shouldn't we just be using `dnf install python3-mypy` so that we are actually using the systems packages here? I see in the test logs that it is still downloading from pypi the packages.

Reply:
Oh, you're only filtering `types-*` there. I was looking at `urllib3-2.2.1` still being downloaded.
```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
requests 2.28.2 requires urllib3<1.27,>=1.21.1, but you have urllib3 2.2.1 which is incompatible.
```

Reply:
> Shouldn't we just be using `dnf install python3-mypy` so that we are actually using the systems packages here? I see in the test logs that it is still downloading from pypi the packages.

I wouldn't think so, because mypy (and pyright) are not tmt requirements, but linters. The idea is to use system dependencies, but defined linters looking at the project from outside. With system mypy, we might be using older versions often, which may not recognize issues newer versions were taught to report.

> Oh, you're only filtering `types-*` there. I was looking at `urllib3-2.2.1` still being downloaded.
> 
> ```
> ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
> requests 2.28.2 requires urllib3<1.27,>=1.21.1, but you have urllib3 2.2.1 which is incompatible.
> ```
> 
> Looking more on the issue of `urllib3`, that seems to be quite a complicated can of worms.

Yeah, it seems to be pulled in because of `types-requests`. TBH, I did not even check the log of `pip install`, and yeah, this is something I'd like to avoid. No idea why **types**-requests feels the need to pull in a genuine package, maybe because `urllib3` does not ship type stubs separately, but an action like this would ruing the idea of "testing against system packages" :/

I think we'll get there, but it is indeed a bigger can of worms than it seemed last week :)

Reply:
> I wouldn't think so, because mypy (and pyright) are not tmt requirements, but linters. The idea is to use system dependencies, but defined linters looking at the project from outside. With system mypy, we might be using older versions often, which may not recognize issues newer versions were taught to report.

Yeah, fully agree. Was more concerned about the actual packages, but then I saw that you are only using `types-*`, which seems a-ok. But I think the version of the `types-*` needs to be matched with the version found in the system's `pip`, i.e. search the `pre-commit-config.yaml` (or maybe other approach later) for the `types-*`packages, search `pip list` or such for the actual package version, and then write the version constrain to match the actual package

> No idea why types-requests feels the need to pull in a genuine package, maybe because urllib3 does not ship type stubs separately, but an action like this would ruing the idea of "testing against system packages" :/

Found out. It's because `urllib3 >= 2.0` already includes typing, and the `types-request` constraint used, does not apply in F40 case, thus getting the latest version `types-request` which uses the latest `urllib3` version. The workflow above should take care of that though, WDYT?